src/engraving/dom/dynamic.cpp

```diff
@@ -942,4 +969,120 @@ String Dynamic::screenReaderInfo() const
     }
     return String(u"%1: %2").arg(EngravingItem::accessibleInfo(), s);
 }
+
+//---------------------------------------------------------
+//   drawEditMode
+//---------------------------------------------------------
+
+void Dynamic::drawEditMode(muse::draw::Painter* p, EditData& ed, double currentViewScaling)
+{
+    if (ed.editTextualProperties) {
+        TextBase::drawEditMode(p, ed, currentViewScaling);
+    } else {
+        EngravingItem::drawEditMode(p, ed, currentViewScaling);
+    }
+}
+
+//---------------------------------------------------------
+//   hasLeftHairpin
+//---------------------------------------------------------
+
+bool Dynamic::hasLeftGrip() const
+{
+    if (segment()->tick().isZero()) {
+        return false; // Don't show the left grip for the leftmost dynamic with tick zero
+    }
+    return m_leftHairpin == nullptr;
+}
+
+//---------------------------------------------------------
+//   hasRightHairpin
+//---------------------------------------------------------
+
+bool Dynamic::hasRightGrip() const
+{
+    return m_rightHairpin == nullptr;
+}
+
+//---------------------------------------------------------
+//   findAdjacentHairpins
+//---------------------------------------------------------
+
+void Dynamic::findAdjacentHairpins()
+{
+    m_leftHairpin = nullptr;
+    m_rightHairpin = nullptr;
+
+    const Fraction tick = segment()->tick();
+    const int intTick = tick.ticks();
+
+    const auto& spanners = score()->spannerMap().findOverlapping(intTick - 1, intTick + 1);
+    for (auto i : spanners) {
+        Spanner* sp = i.value;
+        if (sp->track() == track() && sp->isHairpin()) {
+            Hairpin* hp = toHairpin(sp);
+            if (hp->tick() == tick) {
+                m_rightHairpin = hp;
+            } else if (hp->tick2() == tick) {
+                m_leftHairpin = hp;
+            }
+        }
+    }
+
+    return;
+}
+
+//---------------------------------------------------------
+//   gripsCount
+//---------------------------------------------------------
+
+int Dynamic::gripsCount() const
+{
+    if (empty()) {
+        return 0;
+    }
+
+    if (hasLeftGrip() && hasRightGrip()) {
+        return 2;
+    } else if (hasLeftGrip() ^ hasRightGrip()) {
+        return 1;
+    } else {
+        return 0;
+    }
+}
+
+//---------------------------------------------------------
+//   gripsPositions
+//---------------------------------------------------------
+
+std::vector<PointF> Dynamic::gripsPositions(const EditData&) const
+{
+    const LayoutData* ldata = this->ldata();
+    const PointF pp(pagePos());
+    double md = score()->style().styleS(Sid::hairpinMinDistance).val() * spatium(); // Minimum distance between dynamic and grip
+
+    PointF leftOffset(-ldata->bbox().width() / 2 - md + m_leftDragOffset, -11.408);
+    PointF rightOffset(ldata->bbox().width() / 2 + md + m_rightDragOffset, -11.408);
```

Comment:
The number `-11.408` is what we call a "magic number". When a cryptic hardcoded number is necessary, it is usually a good idea to extract it into a named constant, sometimes accompanied by a comment about how this number was found.
But in this case, I think it should not be a hardcoded number, but computed based on the height of the Dynamic, or perhaps some other parameter like the y-value of the visual center. It should namely also work well with different fonts, font sizes, spatium values.
@mike-spa Could you suggest a good way to compute the vertical grip positions?

Reply:
@cbjeukendrup I calculated it by subtracting the `pagePos` of the dynamic from the position of Grip::START from `HairpinSegment::gripsPositions` and it turned out to be a constant value. I wasn't sure how this could be computed, so I hardcoded it.